flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/runner/ZooKeeperDefaultDispatcherRunnerTest.java

```diff
@@ -166,21 +167,28 @@ public void testResourceCleanupUnderLeadershipChange() throws Exception {
 				defaultDispatcherRunnerFactory)) {
 
 				// initial run
-				DispatcherGateway dispatcherGateway = grantLeadership(dispatcherLeaderElectionService);
+				final DispatcherGateway dispatcherGateway = grantLeadership(dispatcherLeaderElectionService);
 
 				LOG.info("Initial job submission {}.", jobGraph.getJobID());
 				dispatcherGateway.submitJob(jobGraph, TESTING_TIMEOUT).get();
 
+				// wait until job is running
+				CommonTestUtils.waitUntilCondition(() ->
+					dispatcherGateway.requestJobStatus(jobGraph.getJobID(), TESTING_TIMEOUT).get() == JobStatus.RUNNING, Deadline.fromNow(VERIFICATION_TIMEOUT), 20L);
+
 				dispatcherLeaderElectionService.notLeader();
 
 				// recovering submitted jobs
 				LOG.info("Re-grant leadership first time.");
-				dispatcherGateway = grantLeadership(dispatcherLeaderElectionService);
+				DispatcherGateway nextDispatcherGateway = grantLeadership(dispatcherLeaderElectionService);
 
 				LOG.info("Cancel recovered job {}.", jobGraph.getJobID());
 				// cancellation of the job should remove everything
-				final CompletableFuture<JobResult> jobResultFuture = dispatcherGateway.requestJobResult(jobGraph.getJobID(), TESTING_TIMEOUT);
-				dispatcherGateway.cancelJob(jobGraph.getJobID(), TESTING_TIMEOUT).get();
+
+				nextDispatcherGateway.cancelJob(jobGraph.getJobID(), TESTING_TIMEOUT).get();
+				LOG.info("Cancellation successful");
+				// TODO this future was created before the cancellation. Intentionally?
```

Comment:
I don't think it was intentional. It is actually rather questionable, and it would make sense to check how the test managed to pass so far.